packages/config/src/Config.ts

```diff
@@ -1,12 +1,118 @@
 import JsonFile, { JSONObject } from '@expo/json-file';
-import fs from 'fs-extra';
 import path from 'path';
 import slug from 'slugify';
 
-import { AppJSONConfig, ExpRc, ExpoConfig, PackageJSONConfig, ProjectConfig } from './Config.types';
+import {
+  AppJSONConfig,
+  ConfigContext,
+  ExpRc,
+  ExpoConfig,
+  PackageJSONConfig,
+  Platform,
+  ProjectConfig,
+} from './Config.types';
 import { ConfigError } from './Errors';
+import { findAndEvalConfig } from './getConfig';
+import { getRootPackageJsonPath, projectHasModule } from './Modules';
 import { getExpoSDKVersion } from './Project';
 
+/**
+ * Get all platforms that a project is currently capable of running.
+ *
+ * @param projectRoot
+ * @param exp
+ */
+function getSupportedPlatforms(
+  projectRoot: string,
+  exp: Pick<ExpoConfig, 'nodeModulesPath'>
+): Platform[] {
+  const platforms: Platform[] = [];
+  if (projectHasModule('react-native', projectRoot, exp)) {
+    platforms.push('ios', 'android');
+  }
+  if (projectHasModule('react-native-web', projectRoot, exp)) {
+    platforms.push('web');
+  }
+  return platforms;
+}
+
+type GetConfigOptions = {
+  mode: 'development' | 'production';
+  configPath?: string;
+  skipSDKVersionRequirement?: boolean;
+};
+
+function getConfigContext(
+  projectRoot: string,
+  options: GetConfigOptions
+): { context: ConfigContext; pkg: JSONObject } {
+  // TODO(Bacon): This doesn't support changing the location of the package.json
+  const packageJsonPath = getRootPackageJsonPath(projectRoot, {});
+  const pkg = JsonFile.read(packageJsonPath);
+
+  const configPath = options.configPath || customConfigPaths[projectRoot];
+
+  // If the app.json exists, we'll read it and pass it to the app.config.js for further modification
```

Comment:
Maybe we want to get rid of this capability, and make the JSON/JS configurations mutually exclusive, for a few reasons:
- Having the configuration split across two files is more confusing for new users or anyone coming across a project that does this. Now you must know two ways to do the same thing. Now you need to check two files to confirm what configuration value is used.
- When responding to issues the common question "please show us your app.config.js" becomes "please show us your app.config.js and app.json (and package.json)"
- JS can include JSON inside it. It can also import configuration from an arbitrary JSON file, if a static format is desired for some values.

Reply:
> Having the configuration split across two files

We support this functionality for values that are generated by Expo CLI like `bundleIdentifier`. I'm totally open to better ideas though. I imagine a new project's config would look something like:

```js
module.exports = function({ config }) {
  return {
  ...config,
  icon: 'assets/icon.png',
  /* etc... */
  }
}
```

The spread of another config is somewhat unpleasant.

> responding to issues

We can add a mechanism for printing out the serialized config after it's been returned from `app.config.js`.

> if a static format is desired for some values.

I imagine eventually `app.json` will be the static location for generated values.

Reply:
> When responding to issues the common question "please show us your app.config.js" becomes "please show us your app.config.js and app.json (and package.json)"

maybe we could add this to `expo diagnostics`? and we could automatically filter out any potentially sensitive fields

Reply:
all we'd have to do to support this is switch the method used to read the config file.